framework/include/utils/PiecewiseByBlockLambdaFunctor.h

```diff
@@ -0,0 +1,298 @@
+typename PiecewiseByBlockLambdaFunctor<T>::ValueType
+PiecewiseByBlockLambdaFunctor<T>::evaluate(const Elem * const & elem, unsigned int state) const
+{
+  mooseAssert(elem && elem != libMesh::remote_elem,
+              "The element must be non-null and non-remote in functor material properties");
+  auto it = _elem_functor.find(elem->subdomain_id());
+  mooseAssert(it != _elem_functor.end(), subdomainErrorMessage(elem->subdomain_id()));
+  return it->second(elem, state);
+}
+
+template <typename T>
+typename PiecewiseByBlockLambdaFunctor<T>::ValueType
+PiecewiseByBlockLambdaFunctor<T>::evaluate(const ElemFromFaceArg & elem_from_face,
+                                           unsigned int state) const
+{
+  mooseAssert(
+      (std::get<0>(elem_from_face) && std::get<0>(elem_from_face) != libMesh::remote_elem) ||
+          std::get<1>(elem_from_face),
+      "The element must be non-null and non-remote or the face must be non-null in functor "
+      "material properties");
+  auto it = _elem_from_face_functor.find(std::get<2>(elem_from_face));
+  mooseAssert(it != _elem_from_face_functor.end(),
+              subdomainErrorMessage(std::get<2>(elem_from_face)));
+  return it->second(elem_from_face, state);
+}
+
+template <typename T>
+typename PiecewiseByBlockLambdaFunctor<T>::ValueType
+PiecewiseByBlockLambdaFunctor<T>::evaluate(const SingleSidedFaceArg & face,
+                                           unsigned int state) const
+{
+  const auto sub_id = std::get<3>(face);
+  auto it = _face_functor.find(sub_id);
+  mooseAssert(it != _face_functor.end(), subdomainErrorMessage(sub_id));
+  return it->second(face, state);
+}
+
+template <typename T>
+typename PiecewiseByBlockLambdaFunctor<T>::ValueType
+PiecewiseByBlockLambdaFunctor<T>::evaluate(const FaceArg & face, unsigned int state) const
+{
+  using namespace Moose::FV;
+
+  const auto elem_sub_id = std::get<4>(face).first;
+  const auto neighbor_sub_id = std::get<4>(face).second;
+  const auto limiter = std::get<1>(face);
+  const auto * const face_info = std::get<0>(face);
+  mooseAssert(face_info,
+              "We must have a non-null face_info in order to prepare our ElemFromFace tuples");
+  const bool fi_elem_is_upwind = std::get<2>(face);
+  static const typename libMesh::TensorTools::IncrementRank<T>::type example_gradient(0);
+
+  switch (limiter)
+  {
+    case LimiterType::CentralDifference:
+    case LimiterType::Upwind:
+    {
+      const auto elem_from_face = std::make_tuple(&face_info->elem(), face_info, elem_sub_id);
+      const auto neighbor_from_face =
+          std::make_tuple(face_info->neighborPtr(), face_info, neighbor_sub_id);
+      const auto & upwind_elem = fi_elem_is_upwind ? elem_from_face : neighbor_from_face;
+      const auto & downwind_elem = fi_elem_is_upwind ? neighbor_from_face : elem_from_face;
+      auto limiter_obj = Limiter<typename LimiterValueType<T>::value_type>::build(limiter);
+      return interpolate(*limiter_obj,
```

Comment:
Yea, there is no guarantee that the lambda provided by the user forwards evaluation calls to a variable. the lambda could be a constant or it could do something with multiple variables or `Functions`. Limitless options essentially